benchexec/tools/metaval.py

```diff
@@ -108,14 +111,22 @@ def cmdline(self, executable, options, tasks, propertyfile=None, rlimits={}):
                 ).Tool()
 
         if verifierName in self.wrappedTools:
-            with self._in_tool_directory(verifierName) as oldcwd:
-                wrappedOptions = self.wrappedTools[verifierName].cmdline(
-                    self.wrappedTools[verifierName].executable(),
-                    options,
-                    [os.path.relpath(os.path.join(oldcwd, "output/ARG.c"))],
-                    os.path.relpath(os.path.join(oldcwd, propertyfile)),
-                    rlimits,
+            tool = self.wrappedTools[verifierName]
+            assert isinstance(
+                tool, BaseTool2
+            ), "we expect that all wrapped tools extend BaseTool2"
+            wrapped_executable = self.wrappedTools[verifierName].executable(
+                BaseTool2.ToolLocator(
+                    tool_directory=self.TOOL_TO_PATH_MAP[verifierName]
                 )
+            )
+            wrappedOptions = self.wrappedTools[verifierName].cmdline(
+                wrapped_executable,
+                options,
+                [os.path.relpath(os.path.join(os.getcwd(), "output/ARG.c"))],
```

Comment:
This actually always produces `["output/ARG.c"]` because you first join it with the current directory and then make it relative to the current directory again. So this can be simplified.

But more importantly, new tool-info modules require a `Task` instance here.

Reply:
Hmm, this worked alright at the time we drafted the code (I think it was https://github.com/sosy-lab/benchexec/pull/602), what changed inbetween?

Reply:
This was done as part of #592, i.e., the introduction of `BaseTool2`. The changes to MetaVal's tool-info module were based on an intermediate draft version of that API.

If you are not yet familiar with it, I recommend having a look at our [migration guide](https://github.com/sosy-lab/benchexec/blob/master/doc/tool-integration.md#migrating-tool-info-modules-to-new-api) that covers the full set of differences between `BaseTool` and `BaseTool2`.

Reply:
That migration guide can be improved, it doesn't even list the method signatures. It will be quicker reading the actual code.

Reply:
> This actually always produces ["output/ARG.c"] because you first join it with the current directory and then make it relative to the current directory again. So this can be simplified.

Yeah this is probably broken anyway because it actually has to be ["../output/ARG.c"], because the output folder will actually be created outside of the wrapped tool's directory. I guess this is a result of the refactoring where we replace the context manager. At that time the code probably used the cached working directory of metaval instead of `os.getcwd()`. I will need to do some tool testing anyway to see whether this still works.

Reply:
OK, there are still a lot of tool info modules that are not yet refactored to implement`BaseTool2`, so the missing checks for `BaseTool` or `BaseTool2` that you wanted me to remove actually are now hindering me. I basically have to wait for https://github.com/sosy-lab/benchexec/pull/646 to be merged.

Reply:
https://github.com/sosy-lab/benchexec/pull/650/commits/4f1313e86e09889897cb099833e3783265b23b3e should fix this by passing a proper `Task` object, but I have no way of knowing until #646 is merged, I manually merge that PR into this PR (with high chance that this will cause more work than it will save) or readd the `BaseTool` checks

Reply:
Do you mean because you need a working `BaseTool2` tool-info module to test your PR? You can do this with CPAchecker's module, it has been upgraded already.

Reply:
From my point of view, this and the (now 3) unused imports are the only remaining things to do. So please ping me after you finished and tested this, then I will merge.

Reply:
Note that now #646 was merged, so not only CPAchecker's tool-info module but also others are now available for testing.

Reply:
Open points should be addressed via:
https://github.com/sosy-lab/benchexec/pull/650/commits/444a46ce90c679e9f3213d0360a03461afd0add1
https://github.com/sosy-lab/benchexec/pull/650/commits/e9ae4f3dfe583108ecc62286513277e343b1091a
https://github.com/sosy-lab/benchexec/pull/650/commits/8fee3ad6a5fba4ac314ff2bda016b17174a5c993

I also found some bugs via testing and fixed those as well:
https://github.com/sosy-lab/benchexec/pull/650/commits/0bc36059a125f88f91e5c9c9d2e492bc2d693282

Accidentally overlooked/reintroduced some unused imports, those should be fine with:
https://github.com/sosy-lab/benchexec/pull/650/commits/b2adf40319467c20d072c0966b5e2ecce57d8064